vip-jetpack/connection-pilot/class-jetpack-connection-attendant.php

```diff
@@ -0,0 +1,309 @@
+
+		if ( ! method_exists( $jp_connection, 'get_connection_owner' ) ) {
+			return null;
+		}
+
+		$connection_owner = $jp_connection->get_connection_owner();
+		if ( false === $connection_owner ) {
+			return null;
+		}
+
+		return $connection_owner;
+	}
+
+	/**
+	 * Ensures the VIP user exists on the site and has the necessary permissions.
+	 *
+	 * @return WP_User|WP_Error WP_User if successful, WP_Error otherwise.
+	 */
+	public function ensure_user_exists() {
+		$details = $this->get_expected_owner_details();
+
+		$user = get_user_by( 'login', $details['login'] );
+
+		// Create user if it doesn't exist yet.
+		if ( ! $user ) {
+			$result = wp_insert_user( [
+				'user_login'   => $details['login'],
+				'user_email'   => $details['email'],
+				'display_name' => $details['display_name'],
+				'role'         => $details['role'],
+				'user_pass'    => wp_generate_password( 100 ),
+			] );
+
+			if ( is_wp_error( $result ) ) {
+				return new WP_Error( 'jp-cxn-pilot-owner-creation-failed', 'Failed to create new user.' );
+			}
+
+			$user = get_userdata( $result );
+			if ( ! $user ) {
+				return new WP_Error( 'jp-cxn-pilot-owner-creation-failed', 'Failed to create new user.' );
+			}
+		}
+
+		// Ensure the email is correct & hasn't been changed.
+		if ( $details['email'] !== $user->user_email ) {
+			$result = wp_update_user( [
+				'ID'         => $user->ID,
+				'user_email' => $details['email'],
+			] );
+
+			if ( is_wp_error( $result ) ) {
+				return new WP_Error( 'jp-cxn-pilot-owner-update-failed', 'Failed to correct the email address on the primary user.' );
+			}
+		}
+
+		// Add user to blog if needed, and ensure they are a super admin.
+		if ( is_multisite() ) {
+			$blog_id = get_current_blog_id();
+
+			if ( ! is_user_member_of_blog( $user->ID, $blog_id ) ) {
+				$added_to_blog = add_user_to_blog( $blog_id, $user->ID, $details['role'] );
+
+				if ( is_wp_error( $added_to_blog ) ) {
+					return new WP_Error( 'jp-cxn-pilot-owner-ms-add-failed', 'Failed to add user to blog.' );
+				}
+			}
+
+			if ( ! is_super_admin( $user->ID ) ) {
+				// Returns false even if user already is SA.
+				grant_super_admin( $user->ID );
+			}
+		}
+
+		// Ensure the correct role is applied.
+		if ( ! in_array( $details['role'], (array) $user->roles, true ) ) {
+			$user->set_role( $details['role'] );
+		}
+
+		return $user;
+	}
+
+	private function get_expected_owner_details() {
+		$legacy_owner_details = $this->get_legacy_owner_details();
+
+		// NOTE: This filter should be applied very early. Use before plugins_loaded priority 25 to be sure.
+		if ( null !== $legacy_owner_details && ! (bool) apply_filters( 'vip_jetpack_connection_pilot_use_org_owner', false ) ) {
+			return $legacy_owner_details;
+		}
+
+		return $this->get_org_owner_details();
+	}
+
+	private function get_org_owner_details() {
+		$org_id = defined( 'VIP_ORG_ID' ) ? constant( 'VIP_ORG_ID' ) : 0;
+
+		return [
+			'email'        => "vip-jetpack-owner+{$org_id}@wpvip.com",
```

Comment:
Do we really need `vip-` here? since we're already `@wpvip.com`

Reply:
Makes it extra VIP! 😆